### ISSUE TITLE
Train only after doesn't work as DataCollator overwrites labels (incl. -100) with input_ids

### ISSUE THREAD

**Issue by kabachuha** [oobabooga/text-generation-webui#4031]
### Describe the bug

So, in attempt to mask loss for Train only after training training.py sets the labels belonging to before tokens to -100s.

https://github.com/oobabooga/text-generation-webui/blob/ee7bf49804bfb53775424d4c12294445497b610f/modules/training.py#L341-L358

However, the sus thing was that it sets the rest of the labels to 1s — which would be a nonsense. Yet, training happens. Checking it, I found that this still makes sense thanks to the fact that the code uses the DataCollatorForLanguageModeling.

https://github.com/oobabooga/text-generation-webui/blob/ee7bf49804bfb53775424d4c12294445497b610f/modules/training.py#L592

Now, I checked the code for the Collator and found out that it does overwrite all the labels, including the ones having -100 as the values, unless *the labels themselves* are set to `pad_token_id`.

https://github.com/huggingface/transformers/blob/main/src/transformers/data/data_collator.py#L744-L749

```python
        else:
            labels = batch["input_ids"].clone()
            if self.tokenizer.pad_token_id is not None:
                labels[labels == self.tokenizer.pad_token_id] = -100
            batch["labels"] = labels
        return batch
```

So, it's a bug then

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Make a training with or without train only after on

### Logs

```shell
Unrelated
```

### System Info

```shell
V100
```

**Comment by FartyPants** [comment]
Yeah, I think they meant to be a boolean operation where input_ids is pad_id, blank labels with -100 regardless what there is.
